src/Microsoft.AspNetCore.Authentication/VirtualAuthenticationHandler.cs

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    /// <summary>
+    /// Forwards calls to another authentication scheme.
+    /// </summary>
+    public class VirtualAuthenticationHandler : IAuthenticationHandler, IAuthenticationSignInHandler
+    {
+        protected IOptionsMonitor<VirtualSchemeOptions> OptionsMonitor { get; }
+        public AuthenticationScheme Scheme { get; private set; }
+        public VirtualSchemeOptions Options { get; private set; }
+        protected HttpContext Context { get; private set; }
+
+        public VirtualAuthenticationHandler(IOptionsMonitor<VirtualSchemeOptions> options)
+        {
+            OptionsMonitor = options;
+        }
+
+        /// <summary>
+        /// Initialize the handler, resolve the options and validate them.
+        /// </summary>
+        /// <param name="scheme"></param>
+        /// <param name="context"></param>
+        /// <returns>A Task.</returns>
+        public Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
+        {
+            if (scheme == null)
+            {
+                throw new ArgumentNullException(nameof(scheme));
+            }
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            Scheme = scheme;
+            Context = context;
+
+            Options = OptionsMonitor.Get(Scheme.Name) ?? new VirtualSchemeOptions();
+            return Task.CompletedTask;
+        }
+
+        protected virtual string ResolveTarget(string scheme)
+            => scheme ?? Options.DefaultTargetSelector?.Invoke(Context) ?? Options.DefaultTarget;
+
+        public virtual Task<AuthenticateResult> AuthenticateAsync()
+            => Context.AuthenticateAsync(ResolveTarget(Options.AuthenticateTarget));
+
+        public virtual Task SignInAsync(ClaimsPrincipal user, AuthenticationProperties properties)
+            => Context.SignInAsync(ResolveTarget(Options.SignInTarget), user, properties);
+
+        public virtual Task SignOutAsync(AuthenticationProperties properties)
+            => Context.SignOutAsync(ResolveTarget(Options.SignOutTarget), properties);
+
+        public Task ChallengeAsync(AuthenticationProperties properties)
+            => Context.ChallengeAsync(ResolveTarget(Options.ChallengeTarget), properties);
+
+        public Task ForbidAsync(AuthenticationProperties properties)
+            => Context.ForbidAsync(ResolveTarget(Options.ForbidTarget), properties);
```

Comment:
This should follow the same fallback rules as AuthSchemeProvider.
E.g. Forbid -> Challenge -> Default.
https://github.com/aspnet/HttpAbstractions/blob/c0f937239a0a099b73c67c96ab9e1c875952f67f/src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs#L71-L74

Reply:
It's interesting to think that this creates two levels of fallback. VirtualForbid -> VirtualChallenge -> VirtualDefault -> DefaultForbid -> DefaultChallenge -> DefaultDefault.

Reply:
It doesn't need to do the explicit fallbacks since it defaults to null which already captures the other fallbacks.

Reply:
That was the whole point of default: null, external callers to Context's auth apis will have the default logic taken care of already.

Reply:
So: VirtualForbid -> VirtualDefault -> DefaultForbid -> DefaultChallenge -> DefaultDefault.
It seems strange to leave VirtualChallenge out of that chain when VirtualDefault is included.

Reply:
I don't think we want this to mimic any of the crazy fallback chains, this is a purely virtual scheme. There should be a main default which you can easily set so you don't have to set every property.  There should be one which takes precedence for each action.  

But if we do find a need for the extra fallbacks in the virtual schemes in the upcoming sample/scenarios, we can then bring them back as needed